sosetup: Production Mode should auto-configure PF_RING instances #735

weslambert opened the pull request:
**** A summary of changes ****
*** In reference to issue #735***

When running sosetup, choosing "Production Mode", then "Best Practices":

IDS/Bro processes are automatically configured based on the number of cores.

**IDS**: $CORES (number of cores)
**Bro**: $HALF_CORES (half the number of cores)

When running sosetup, choosing "Production Mode", then "Custom":

**IDS Engine Procs**

How many IDS engine processes would you like to run?

Based on your system's number of cores, it is recommended that you run $CORES IDS engine processes per interface.

If you need to change this setting later, change IDS_LB_PROCS in /etc/nsm/HOSTNAME-INTERFACE/sensor.conf"

(User selects number of cores)

**BRO_LB_PROCS**
How many Bro processes would you like to run?

This is limited by the total number of CPU cores on your system,
but you should probably choose no more than HALF ($HALF_CORES) your number of CPU cores ($CORES).

(User selects number of cores)

Thanks,
Wes

dougburks commented:
Hi Wes,

Thanks for the pull request!

Bottom line up front: I think we should go ahead and change IDS to be $HALF_CORES just like Bro to avoid overworking the box.

More detail:

From your summary, it sounds like you were faithful to the wording that already exists in Setup.  However, I think we should probably modify the recommendation and, in the case of "Best Practices", the value that gets automatically set.  

We need to think about dedicating CPU cores to processes so that at some point in the future we can set CPU affinity and keep processes bound to CPU cores and be the only process on the core.

CPU cores should be dedicated as follows:
1 for OS itself
1 for netsniff-ng
remaining cores split between IDS and Bro (in the case of "Best Practices")

For example, suppose you have 12 CPU cores:
1 for OS itself
1 for netsniff-ng
5 for IDS
5 for Bro

If the user chooses Custom, those additional processes may require additional CPUs, but I think the above may give a good starting point for most users.

Thoughts?

weslambert commented:
Doug, 

Thanks for the feedback. 

So, I'm assuming, for a sensor or standalone, a machine should have a bare minimum number of 4 cores?  if a user tries to run Security Onion on a machine with 2 cores, should this just default to dividing cores between IDS/Bro or would this not be supported?  

Base on my assumption from the recommendation(s) above, we would be looking for something like the following:
 
$CORES could equal the number of CPU cores
$SO_CORES (or similarly named variable) would equal the number of CPU cores ($CORES) available minus 2.  
$SO_CORES could then be split in half ($HALF_CORES) for both the IDS and BRO processes.

Is this a correct assumption?

Thanks,
Wes

dougburks commented:
If the user has less than 4 CPU cores, then I would just default to IDS_LB_PROCS=1 and BRO_LB_PROCS=1.

You're on the right track, but we also need to take into account how many sniffing interfaces there are.

For example:
2 sniffing interfaces (eth1 and eth2)
12 CPU cores
minus 1 core for the OS itself
equals 11 CPU cores remaining for sniffing processes
divided by 2 sniffing interfaces
equals 5.5 CPU cores for each sniffing interface
minus 1 CPU core for netsniff-ng
leaves 4.5 (round down to 4) CPU cores to be divided between IDS and BRO

IDS_LB_PROCS=2
BRO_LB_PROCS=2

results in:
2 IDS processes on eth1 (2 CPU cores)
2 IDS processes on eth2 (2 CPU cores)
2 BRO workers on eth1 (2 CPU cores)
2 BRO workers on eth2 (2 CPU cores)
1 netsniff-ng on eth1 (1 CPU core)
1 netsniff-ng on eth2 (1 CPU core)
OS itself (1 CPU core)
11 CPU cores assigned

weslambert commented:
Thanks for the clarification, Doug.  I've changed sosetup to accommodate your suggestions as follows:

The basic formula I have is as follows:

(((CORES - 1 ) / SNIFF_INTERFACES) -1) / 2))

OR

CORES =`grep -c ^processor /proc/cpuinfo`
SO_CORES=$(((CORES - 1) / SNIFF INTERFACES))
CALCD_CORES=$(((SO_CORES -1 ) / 2))

SNIFF_INTERFACES=`awk '/manual/ {print $2}' /etc/network/interfaces | wc -l`

SO_CORES= The number of machine cores, minus the reserved core for the OS, divided by the number of sniffing interfaces,

CALCD CORES= Number of process cores per interface, minus reserved core for netsniff-ng (per process), split in half to accomodate IDS/Bro.

Please let me know your thoughts.

Thanks,
Wes
